sdk/core/test/integration/otp/index.test.ts

```diff
@@ -1,417 +1,426 @@
+    const [email, tag, messageParameters] = prepareOtpMessageParameters(testId)
 
-    await wait(DELAY)
-    const otpNew = await getOtp()
+    const timestamp = Date.now()
+    await commonNetworkMember.changeUsername(email, options, messageParameters)
+    const otpCode = await waitForOtpCode(tag, timestamp)
 
-    const commonNetworkMemberNew = await CommonNetworkMember.confirmSignUp(tokenNew, otpNew, options)
+    await commonNetworkMember.confirmChangeUsername(email, otpCode, options)
+    await commonNetworkMember.signOut()
 
-    expect(commonNetworkMemberNew).to.exist
+    commonNetworkMember = await CommonNetworkMember.fromLoginAndPassword(email, password, options)
+    expect(commonNetworkMember).to.be.an.instanceof(CommonNetworkMember)
 
-    let responseErrorNew
+    const newPassword = `${password}_updated`
 
-    try {
-      await commonNetworkMemberNew.changeUsername(cognitoUsername, options)
-    } catch (error) {
-      responseErrorNew = error
-    }
+    await commonNetworkMember.changePassword(password, newPassword, options)
+    await commonNetworkMember.signOut()
 
-    expect(responseErrorNew).to.exist
-    expect(responseErrorNew.name).to.eql('COR-7')
+    commonNetworkMember = await CommonNetworkMember.fromLoginAndPassword(email, newPassword, options)
+    expect(commonNetworkMember).to.be.an.instanceof(CommonNetworkMember)
   })
 
-  it('#signUp with username, add email, signIn with email, change password', async () => {
-    const cognitoUsername = generateUsername()
-
-    let networkMember = await CommonNetworkMember.signUp(cognitoUsername, cognitoPassword, options)
-
-    expect(networkMember).to.be.an.instanceof(CommonNetworkMember)
-
-    const email = generateEmail()
-
-    await networkMember.changeUsername(email, options)
-
-    await wait(DELAY)
-    const otp = await getOtp()
-
-    await networkMember.confirmChangeUsername(email, otp, options)
-
-    await networkMember.signOut()
-
-    networkMember = await CommonNetworkMember.fromLoginAndPassword(email, cognitoPassword, options)
-
-    expect(networkMember).to.be.an.instanceof(CommonNetworkMember)
-
-    const newPassword = generateUsername() // test.user-175AB... - is OKAY
-
-    await networkMember.changePassword(cognitoPassword, newPassword, options)
-
-    await networkMember.signOut()
-
-    networkMember = await CommonNetworkMember.fromLoginAndPassword(email, newPassword, options)
-
-    expect(networkMember).to.be.an.instanceof(CommonNetworkMember)
+  describe('[with existing user]', () => {
+    let email: string
+    let tag: string
+    let messageParameters: MessageParameters
+
+    // reuse the same user for all subsequent tests to reduce OTP emails count
+    before(async () => {
+      // testId is not available in "before" hook, so we generate it separately
+      ;[email, tag, messageParameters] = prepareOtpMessageParameters(generateTestId())
```

Comment:
Not sure. I also thought about it. Perhaps generating it once (where possible, like now) reduces the risk of collision even more.
It's not that important, though, since emails are only "waited" since specified timestamp.
Subjectively it just looks somewhat "cleaner" when we have an id for a test run instead of generating random string each time we need new email. And it also corresponds to the documentation of testmail.

Reply:
Then, since it is synchronous, maybe it is better to move it out of the top-level `describe` and just do `const testId = generateTestId()` at the top of the module? That way all tests will be able to use it; you won't have to think whether it is available in "before" hook or not.

Reply:
Sorry for the confusion. The line we're currently discussing is an exception (since we reuse the user account).
"testId" should be unique in each spec. Though, now that I've said it, it would be easier to generate new id every time we generate an email.
I will remove it.